internal/server/providers/oidc_test.go

```diff
@@ -2,16 +2,850 @@ package providers
+				assert.Equal(t, email, "hello@example.com")
+			},
+		},
+	}
+
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			server.tokenResponse = test.tokenResponse(t)
+			accToken, refToken, accTokenExp, email, err := test.provider.ExchangeAuthCodeForProviderTokens(ctx, "some-auth-code")
+			test.verifyFunc(t, accToken, refToken, accTokenExp, email, err)
+		})
+	}
+}
+
+func TestRefreshAccessToken(t *testing.T) {
+	server, ctx := setupOIDCTest(t)
+	serverURL := server.run(t)
+	provider := NewOIDC(models.Provider{Kind: models.OIDCKind, URL: serverURL, ClientID: "whatever"}, "secret", "http://localhost:8301")
+
+	now := time.Now().UTC()
+
+	claims := jwt.Claims{
+		Audience:  jwt.Audience([]string{server.knownClientID}),
+		NotBefore: jwt.NewNumericDate(now.Add(time.Minute * -5)), // adjust for clock drift
+		Expiry:    jwt.NewNumericDate(now.Add(time.Minute * 5)),
+		IssuedAt:  jwt.NewNumericDate(now),
+		Issuer:    serverURL,
+	}
+
+	body, err := testTokenResponse(claims, server.signingKey, "hello@example.com")
+	assert.NilError(t, err)
+
+	tests := []struct {
+		name          string
+		providerUser  *models.ProviderUser
+		tokenResponse tokenResponse
+		verifyFunc    func(t *testing.T, accessToken string, expiry *time.Time, err error)
+	}{
+		{
+			name: "invalid/expired refresh token fails",
+			providerUser: &models.ProviderUser{
+				AccessToken:  models.EncryptedAtRest("aaa"),
+				RefreshToken: models.EncryptedAtRest("bbb"),
+				ExpiresAt:    time.Now().Add(time.Minute * -5),
+			},
+			tokenResponse: tokenResponse{
+				code: 403,
+				body: "",
+			},
+			verifyFunc: func(t *testing.T, accessToken string, expiry *time.Time, err error) {
+				assert.ErrorContains(t, err, "cannot fetch token")
+			},
+		},
+		{
+			name: "valid access token is not refreshed",
+			providerUser: &models.ProviderUser{
+				AccessToken:  models.EncryptedAtRest("aaa"),
+				RefreshToken: models.EncryptedAtRest("bbb"),
+				ExpiresAt:    time.Now().Add(time.Minute * 5),
+			},
+			verifyFunc: func(t *testing.T, accessToken string, expiry *time.Time, err error) {
+				assert.NilError(t, err)
+				assert.Equal(t, accessToken, "aaa")
+			},
+		},
+		{
+			name: "expired access token is refreshed",
+			providerUser: &models.ProviderUser{
+				AccessToken:  models.EncryptedAtRest("aaa"),
+				RefreshToken: models.EncryptedAtRest("bbb"),
+				ExpiresAt:    time.Now().Add(time.Minute * -5),
+			},
+			tokenResponse: tokenResponse{
+				code: 200,
+				body: body,
+			},
+			verifyFunc: func(t *testing.T, accessToken string, expiry *time.Time, err error) {
+				assert.NilError(t, err)
+				assert.Assert(t, accessToken != "aaa")
+			},
+		},
+	}
+
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			server.tokenResponse = test.tokenResponse
+			accessToken, exp, err := provider.RefreshAccessToken(ctx, test.providerUser)
+			test.verifyFunc(t, accessToken, exp, err)
+		})
+	}
+}
+
+func TestSyncProviderUser(t *testing.T) {
```

Comment:
```suggestion
func TestOIDC_SyncProviderUser(t *testing.T) {
```

This case is specific to the OIDC implementation, is that right, where as the other one is for azure